src/main/java/com/datapipe/jenkins/vault/credentials/snapshots/VaultStringCredentialSnapshotTaker.java

```diff
@@ -0,0 +1,21 @@
+package com.datapipe.jenkins.vault.credentials.snapshots;
+
+import com.cloudbees.plugins.credentials.CredentialsSnapshotTaker;
+import com.datapipe.jenkins.vault.credentials.SecretSnapshot;
+import com.datapipe.jenkins.vault.credentials.common.VaultStringCredential;
+import com.datapipe.jenkins.vault.credentials.common.VaultStringCredentialImpl;
+
+public class VaultStringCredentialSnapshotTaker extends CredentialsSnapshotTaker<VaultStringCredential> {
+
+    @Override
+    public Class<VaultStringCredential> type() {
+        return VaultStringCredential.class;
+    }
+
+    @Override
+    public VaultStringCredential snapshot(
+        VaultStringCredential credential) {
+        SecretSnapshot snapshot = new SecretSnapshot(credential.getSecret());
+        return new VaultStringCredentialImpl(credential.getScope(), credential.getId(), credential.getDescription(), snapshot);
```

Comment:
This does not make any sense to me. Just return a https://javadoc.jenkins.io/plugin/plain-credentials/org/jenkinsci/plugins/plaincredentials/impl/StringCredentialsImpl.html with the actual secret. Ditto for all the other impls. Without knowing much about this plugin I suspect this PR could be way smaller.

Reply:
I don't see how that can be achieved given they type constraints of the [CredentialsSnapshotTaker](https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/java/com/cloudbees/plugins/credentials/CredentialsSnapshotTaker.java) class, the definition of the `snapshot` function is 
```
public abstract C snapshot(C credentials);
```
So you can't convert from a `VaultStringCredentialImpl` to a `StringCredentialsImpl` because they are separate classes that independently derive from `org.jenkinsci.plugins.plaincredentials.StringCredentials`.

Reply:
Ok, I'll see what happens if I add snapshot takers for the basic ssh and password types. I would definitely prefer avoiding large scale architectural changes at this point.

Reply:
> understand how the `CredentialsProvider.snapshot` is supposed to find the right class because I can just get the `writeReplace` to call the correct snapshot taker, which is probably bad for extensibility

Did you have a `CredentialsSnapshotTaker` impl marked with `@Extension`? @jtnord might be able to help.

Reply:
Jesse is on the money as normal https://github.com/jenkinsci/hashicorp-vault-plugin/blob/fa6fb1f215f493900b4091b8dcb0affde3c95c89/src/main/java/com/datapipe/jenkins/vault/credentials/snapshots/VaultSSHUserPrivateKeyCredentialsSnapshotTaker.java#L15 is missing an `Extension` annotation

Reply:
> Did you have a `CredentialsSnapshotTaker` impl marked with `@Extension`?

I did not. That fixed the snapshot taker not being picked up by `CredentialsProvider.snapshot` the passphrase issue remains.

Reply:
Not sure offhand what would be up with the passphrase; sounds like something to track down in a debugger.

As an aside, I see little reason for SSH private keys kept by a CI system to employ a passphrase: the passphrase is just kept alongside the encrypted key, so it adds no real defense. Merely makes it easier for someone to upload a key created originally for personal use with a passphrase. I suppose `ssh-credentials` could just convert it to the raw key upon storing, though IIRC that requires utilities from BouncyCastle.

Reply:
Looking at the SnapshotTaker it does not appear that there is anything specific to vault in the snapshot taker (as in effect you are just making 3 calls to resolve the secrets, which a generic provider would do with a call to getPassphrase(), `getPrivateKey()` and `getUserName()` so a generic snapshot taker that just worked on [`SSHUserPrivateKey`](https://github.com/jenkinsci/ssh-credentials-plugin/blob/95c2fec1c0478c04192f0b9b90496ddc8f3e165e/src/main/java/com/cloudbees/jenkins/plugins/sshcredentials/SSHUserPrivateKey.java)  could work here (and be better off in the `ssh-credentials plugin`)  the reason to have a provider specific implementation is if something needs to be optimized (for example you can make one call to vault to get all the required infor rather than 3 separate ones).

I can not see anything obviously wrong at this point - check what the snapshotter is snapshotting I guess.

Reply:
ok, taking the writeReplace out of `VaultSSHUserPrivateKeyImpl` and having the snapshot taker in `BasicSSHUserPrivateKey`  gets the snapshot taken and passed to the agent, so I think that part can be considered fixed. Thanks for the help.
I'll see what I can work out with git and the ssh keys.

Reply:
> Looking at the SnapshotTaker it does not appear that there is anything specific to vault in the snapshot taker (as in effect you are just making 3 calls to resolve the secrets, which a generic provider would do with a call to getPassphrase(), `getPrivateKey()` and `getUserName()` so a generic snapshot taker that just worked on [`SSHUserPrivateKey`](https://github.com/jenkinsci/ssh-credentials-plugin/blob/95c2fec1c0478c04192f0b9b90496ddc8f3e165e/src/main/java/com/cloudbees/jenkins/plugins/sshcredentials/SSHUserPrivateKey.java) could work here (and be better off in the `ssh-credentials plugin`) the reason to have a provider specific implementation is if something needs to be optimized (for example you can make one call to vault to get all the required infor rather than 3 separate ones).

Well there was such an implementation but it got nuked as part of [SECURITY-440](https://www.jenkins.io/security/advisory/2018-06-25/#SECURITY-440)